ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/RaoUtil.java

```diff
@@ -160,4 +168,52 @@ public static ObjectiveFunctionEvaluator createObjectiveFunction(RaoParameters r
                 throw new NotImplementedException("Not implemented objective function");
         }
     }
+
+    public static List<Optional<Country>> getCnecLocation(BranchCnec cnec, Network network) {
+        return getNetworkElementLocation(cnec.getNetworkElement(), network);
+    }
+
+    private static List<Optional<Country>> getNetworkElementLocation(NetworkElement networkElement, Network network) {
+        Identifiable<?> ne = network.getIdentifiable(networkElement.getId());
+        if (ne instanceof Line) {
+            Line line = (Line) ne;
+            return Arrays.asList(line.getTerminal1().getVoltageLevel().getSubstation().getCountry(), line.getTerminal2().getVoltageLevel().getSubstation().getCountry());
+        } else if (ne instanceof TwoWindingsTransformer) {
+            TwoWindingsTransformer transformer = (TwoWindingsTransformer) ne;
+            return Arrays.asList(transformer.getTerminal1().getVoltageLevel().getSubstation().getCountry(), transformer.getTerminal2().getVoltageLevel().getSubstation().getCountry());
+        } else if (ne instanceof Switch) {
+            return Arrays.asList(((Switch) ne).getVoltageLevel().getSubstation().getCountry());
+        } else {
+            throw new NotImplementedException("Don't know how to figure out the location of " + ne.getId() + " of type " + ne.getClass());
+        }
+    }
+
+    public static List<Optional<Country>> getNetworkActionLocation(NetworkAction networkAction, Network network) {
+        List<List<Optional<Country>>> listOfLists = networkAction.getNetworkElements().stream()
+                .map(networkElement -> getNetworkElementLocation(networkElement, network))
+                .collect(Collectors.toList());
+        List<Optional<Country>> result = new ArrayList<>();
+        listOfLists.forEach(result::addAll);
+        return result;
+    }
+
+    public static BranchCnec getMostLimitingElement(Set<BranchCnec> cnecs, String variantId, Unit unit, boolean relativePositiveMargins) {
+        List<BranchCnec> sortedCnecs = cnecs.stream().
+                filter(BranchCnec::isOptimized).
+                sorted(Comparator.comparingDouble(cnec -> computeCnecMargin(cnec, variantId, unit, relativePositiveMargins))).
```

Comment:
You could use : `map(cnec -> computeCnecMargin(...)).min(Double::compareTo)` instead.

I guess it's quicker to identify the min than to sort all the stream.

Reply:
but we have to get the BranchCnec with the minimum margin, not just the minimum margin
you'd then filter the cnecs to find the one that has a margin equal to the minimum ? (we'd have to recompute margins or store them)

Reply:
Indeed, the min doesn't work then. Forgot my comment ^^